snappy/overlord.go

```diff
@@ -0,0 +1,89 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snappy
+
+import (
+	"path/filepath"
+
+	"github.com/ubuntu-core/snappy/dirs"
+	"github.com/ubuntu-core/snappy/progress"
+)
+
+// Overlord is responsible for the overall system state
+type Overlord struct {
+}
+
+// Install installs the given snap file to the systems.
+//
+// It returns the local snap file or an error
+func (o *Overlord) Install(snapFilePath string, origin string, inter progress.Meter, flags InstallFlags) (*SnapPart, error) {
+	return nil, ErrNotImplemented
+}
+
+// Uninstall remove the given local snap from the system
+//
+// It returns an error on failure
+func (o *Overlord) Uninstall(sp *SnapPart, pb progress.Meter) error {
+	return ErrNotImplemented
+}
+
+// SetActive sets the active state of the given snap
+//
+// It returns an error on failure
+func (o *Overlord) SetActive(sp *SnapPart, active bool, pb progress.Meter) error {
+	return ErrNotImplemented
+}
+
+// Installed returns the installed snaps from this repository
+func (o *Overlord) Installed() (parts []*SnapPart) {
+	globExpr := filepath.Join(dirs.SnapSnapsDir, "*", "*", "meta", "package.yaml")
+	if newParts, err := o.partsForGlobExpr(globExpr); err == nil {
```

Comment:
Such an `Err` method would mean having to remember what can fail or not, and on the way to that making it really hard to review when an error check is being missed. Let's please not go there.

As a rule of thumb functions that can fail should really return errors, in the vast majority of cases. Otherwise we're exchanging coding convenience by stability (exceptions FTW!).

`Installed` will eventually drop that error result, because we're hopefully going to move into an in-memory state model, but I'd strongly recommend having that error result being explicitly returned by now. If something goes wrong, it's much better to be told what went wrong than having an empty list of snaps. It's trivial to drop the error in the call site if that's really irrelevant in some particular circumstance. We cannot tell what the case is from here.